File: benchmarks/beginner/B2-file-line-counter/go/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"time"
)

const LINES = 1_000_000
const RUNS = 10

var dataDir string
var filePath string

func init() {
	dir, _ := filepath.Abs(filepath.Join(".", "..", "..", "..", "..", "data"))
	dataDir = dir
	filePath = filepath.Join(dataDir, "b2_lines.txt")
}

func writeFile() {
	os.MkdirAll(dataDir, 0755)
	f, err := os.Create(filePath)
	if err != nil {
		panic(err)
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for i := 1; i <= LINES; i++ {
		fmt.Fprintf(w, "Line %d: The quick brown fox jumps over the lazy dog\n", i)
	}
	w.Flush()
}

func readAndCount() int {
	f, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	count := 0
	for scanner.Scan() {
		count++
	}
	return count
}

func main() {
	writeTimings := make([]float64, RUNS)
	readTimings := make([]float64, RUNS)

	fmt.Println("=== BENCHMARK: B2 - File Line Counter ===")
	fmt.Println("Language: Go")
	fmt.Printf("Runtime:  Go %s\n", runtime.Version())
	fmt.Println()

	var count int
	for i := 0; i < RUNS; i++ {
		start := time.Now()
		writeFile()
		writeElapsed := float64(time.Since(start).Microseconds()) / 1000.0

		start = time.Now()
		count = readAndCount()
		readElapsed := float64(time.Since(start).Microseconds()) / 1000.0

		writeTimings[i] = writeElapsed
		readTimings[i] = readElapsed
		total := writeElapsed + readElapsed
		fmt.Printf("Run %2d: Write %.2f ms | Read %.2f ms | Total %.2f ms\n", i+1, writeElapsed, readElapsed, total)
	}

	fmt.Println()
	for _, pair := range [][2]interface{}{{"Write", writeTimings}, {"Read", readTimings}} {
		label := pair[0].(string)
		timings := pair[1].([]float64)
		minT, maxT, sum := timings[0], timings[0], 0.0
		for _, t := range timings {
			sum += t
			if t < minT { minT = t }
			if t > maxT { maxT = t }
		}
		avg := sum / float64(RUNS)
		variance := 0.0
		for _, t := range timings {
			variance += (t - avg) * (t - avg)
		}
		stddev := math.Sqrt(variance / float64(RUNS))
		fmt.Printf("%s — Min: %.2f ms | Avg: %.2f ms | Max: %.2f ms | StdDev: %.2f ms\n", label, minT, avg, maxT, stddev)
	}

	totals := make([]float64, RUNS)
	for i := range totals {
		totals[i] = writeTimings[i] + readTimings[i]
	}
	minT, maxT, sum := totals[0], totals[0], 0.0
	for _, t := range totals {
		sum += t
		if t < minT { minT = t }
		if t > maxT { maxT = t }
	}
	avg := sum / float64(RUNS)
	variance := 0.0
	for _, t := range totals {
		variance += (t - avg) * (t - avg)
	}
	stddev := math.Sqrt(variance / float64(RUNS))
	fmt.Printf("Total — Min: %.2f ms | Avg: %.2f ms | Max: %.2f ms | StdDev: %.2f ms\n", minT, avg, maxT, stddev)

	fmt.Println()
	status := "✗"
	if count == LINES {
		status = "✓"
	}
	fmt.Printf("Verification: %d %s\n", count, status)

	os.Remove(filePath)
}

File: benchmarks/beginner/B2-file-line-counter/python/main.py
import os
import time
import math

DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", "data")
FILE_PATH = os.path.join(DATA_DIR, "b2_lines.txt")
LINES = 1_000_000
RUNS = 10


def write_file():
    os.makedirs(DATA_DIR, exist_ok=True)
    with open(FILE_PATH, "w") as f:
        for i in range(1, LINES + 1):
            f.write(f"Line {i}: The quick brown fox jumps over the lazy dog\n")


def read_and_count():
    count = 0
    with open(FILE_PATH, "r") as f:
        for _ in f:
            count += 1
    return count


def main():
    write_timings = []
    read_timings = []

    print("=== BENCHMARK: B2 - File Line Counter ===")
    print("Language: Python")
    print(f"Runtime:  Python {__import__('sys').version.split()[0]}")
    print()

    for i in range(1, RUNS + 1):
        start = time.perf_counter()
        write_file()
        write_elapsed = (time.perf_counter() - start) * 1000

        start = time.perf_counter()
        count = read_and_count()
        read_elapsed = (time.perf_counter() - start) * 1000

        write_timings.append(write_elapsed)
        read_timings.append(read_elapsed)
        total = write_elapsed + read_elapsed
        print(f"Run {i:2d}: Write {write_elapsed:.2f} ms | Read {read_elapsed:.2f} ms | Total {total:.2f} ms")

    print()
    for label, timings in [("Write", write_timings), ("Read", read_timings)]:
        min_t = min(timings)
        avg_t = sum(timings) / len(timings)
        max_t = max(timings)
        variance = sum((t - avg_t) ** 2 for t in timings) / len(timings)
        stddev = math.sqrt(variance)
        print(f"{label} — Min: {min_t:.2f} ms | Avg: {avg_t:.2f} ms | Max: {max_t:.2f} ms | StdDev: {stddev:.2f} ms")

    totals = [w + r for w, r in zip(write_timings, read_timings)]
    min_t = min(totals)
    avg_t = sum(totals) / len(totals)
    max_t = max(totals)
    variance = sum((t - avg_t) ** 2 for t in totals) / len(totals)
    stddev = math.sqrt(variance)
    print(f"Total — Min: {min_t:.2f} ms | Avg: {avg_t:.2f} ms | Max: {max_t:.2f} ms | StdDev: {stddev:.2f} ms")

    print()
    status = "✓" if count == LINES else "✗"
    print(f"Verification: {count} {status}")

    os.remove(FILE_PATH)


if __name__ == "__main__":
    main()
